lib/src/rules/equals_test_argument_type.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.equals_test_argument_type;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/linter.dart';
+
+const desc = r'"operator ==(Object other)" should test argument type.';
+
+const details = r'''
+**AVOID** "operator ==(Object other)" should test argument type.
+Not doing so might result in Null Pointer Exceptions which will be unexpected
+for consumers of your class.
+
+**GOOD:**
+```
+class Field {
+}
+
+class Good {
+  final Field someField;
+
+  Good(this.someField);
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    return other is Good &&
+        this.someField == other.someField;
+  }
+
+  @override
+  int get hashCode {
+    return someField.hashCode;
+  }
+}
+```
+
+**BAD:**
+```
+class Field {
+}
+
+class Bad {
+  final Field someField;
+
+  Bad(this.someField);
+
+  @override
+  bool operator ==(Object other) {
+    Bad otherBad = other as Bad; // LINT
+    bool areEqual = otherBad != null && otherBad.someField == someField;
+    return areEqual;
+  }
+
+  @override
+  int get hashCode {
+    return someField.hashCode;
+  }
+}
+```
+''';
+
+class EqualsTestArgumentType extends LintRule {
+  EqualsTestArgumentType()
+      : super(
+            name: 'equals_test_argument_type',
+            description: desc,
+            details: details,
+            group: Group.errors);
+
+  @override
+  AstVisitor getVisitor() => new _Visitor(this);
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  visitAsExpression(AsExpression node) {
+    MethodDeclaration declaration =
+        node.getAncestor((n) => n is MethodDeclaration);
+    if (!_isEqualsOverride(declaration) ||
+        node.expression is! SimpleIdentifier) {
+      return;
+    }
+
+    SimpleIdentifier identifier = node.expression;
+    String parameterName = declaration.parameters.parameterElements.first.name;
```

Comment:
Method declarations may have `null` parameters in which case I think we want to be safe and guard with `?.`.  Moreover, if `parameterElements` haven't been resolved, `first` may contain a `null`? In any case, I think playing safe wouldn't hurt.

```String parameterName = declaration.parameters?.parameterElements?.first?.name;```

?.yuck. :)

@bwilkerson : thoughts?